crates/wasm-mutate/src/mutators/peephole/eggsy/encoder.rs

```diff
@@ -28,58 +28,23 @@ enum TraversalEvent {
 }
 
 impl Encoder {
-    pub(crate) fn expr2wasm(
+    /// Infers types for the rewriting egraph tree
+    /// During the translation from Wasm to the egraph some type information is missin
+    /// However, it can be inferred by the type relations between the nodes, for example
+    /// for `i32.wrap_i64` operator, we always know that it returns an i32 constant and waits
+    /// for the operand to be an i64 integer. This information could be propagated to the other
+    /// nodes in the tree. This function does exactly that, it fixes some nodes to expect
+    /// and receive specific types using fixed operators and function signatures. The type tree
+    /// is traversed until is converges, if for some reason a node has not a type (when it should)
+    /// this type is gathered from the egraph eclass data.
+    fn infer_types(
         info: &ModuleInfo,
-        rnd: &mut rand::prelude::SmallRng,
         expr: &RecExpr<Lang>,
-        node_to_eclass: &Vec<Id>,
-        newfunc: &mut Function,
+        node_to_eclass: &[Id],
         egraph: &EG,
-    ) -> crate::Result<()> {
-        // This is a patch, this logic should be here
-        // If root is Unfold...bypass
+    ) -> crate::Result<Vec<Option<PrimitiveTypeInfo>>> {
```

Comment:
We can always find the type of an expression, right? If so, then we should return `Result<Vec<PrimititiveTypeInfo>>` without the `Option` in there.

Reply:
Not necessarily, some nodes are just helpers, like for example `Arg` nodes, we always should know how to write them down, but the type information is not needed for them when they are in the right operator.

On the other hand, if for some reason they are moved out of the corresponding operator, for example, the `memory-pop-out` mutator (the only case so far), its type is inferred by the fixed point since it corresponds to another operator then.